storage/jellyfish-merkle/src/lib.rs

```diff
@@ -322,64 +373,43 @@ where
     /// the returned batch, the state `S_{i+1}` is ready to be read from the tree by calling
     /// [`get_with_proof`](struct.JellyfishMerkleTree.html#method.get_with_proof). Anything inside
     /// the batch is not reachable from public interfaces before being committed.
-    pub fn batch_put_value_sets(
+    pub fn batch_put_value_set(
         &self,
-        value_sets: Vec<Vec<(HashValue, &(HashValue, K))>>,
-        node_hashes: Option<Vec<&HashMap<NibblePath, HashValue>>>,
+        value_set: Vec<(HashValue, &(HashValue, K))>,
+        node_hashes: Option<&HashMap<NibblePath, HashValue>>,
         persisted_version: Option<Version>,
-        first_version: Version,
-    ) -> Result<(Vec<HashValue>, TreeUpdateBatch<K>)> {
-        let mut tree_cache = TreeCache::new(self.reader, first_version, persisted_version)?;
-        let hash_sets: Vec<_> = match node_hashes {
-            Some(hashes) => hashes.into_iter().map(Some).collect(),
-            None => (0..value_sets.len()).map(|_| None).collect(),
-        };
-
-        for (idx, (value_set, hash_set)) in
-            itertools::zip_eq(value_sets.into_iter(), hash_sets.into_iter()).enumerate()
-        {
-            if value_set.is_empty() {
-                tree_cache.freeze();
-                continue;
-            }
-
-            let version = first_version + idx as u64;
-            let deduped_and_sorted_kvs = value_set
-                .into_iter()
-                .collect::<BTreeMap<_, _>>()
-                .into_iter()
-                .collect::<Vec<_>>();
-            let root_node_key = tree_cache.get_root_node_key().clone();
-            let (new_root_node_key, _) = self.batch_insert_at(
-                root_node_key,
+        version: Version,
+    ) -> Result<(HashValue, TreeUpdateBatch<K>)> {
+        let deduped_and_sorted_kvs = value_set
```

Comment:
do we still need this step?

Reply:
Looks like so. At least there's no guarantee in the interface that the input value set is unique and sorted. 
I always wanted to introduce a `struct SortedUniqueKvs` to make it clear.